### PR TITLE
DHFPROD-3347: DEADLOCK/performance changes

### PR DESCRIPTION
**Pull request by ryanjdew**
See https://project.marklogic.com/jira/browse/DHFPROD-3347 and https://project.marklogic.com/jira/browse/DHFPROD-3425 for more information on why these changes are needed.

### CODE REVIEW

**Comment by aebadirad** [comment]
Holy moly. Can we get something written about this in the wiki so we can track the shift to document sources grouped by URI and the sub-update-transactions we're now implementing?

File: marklogic-data-hub/src/main/resources/ml-modules/root/com.marklogic.smart-mastering/survivorship/merging/base.xqy [hunk 1]
```diff
@@ -2357,45 +2354,52 @@ declare variable $locked-uris-map as map:map := map:map();
...
-          (: We only care about locks by mastering. We'll still need to wait on outside transactions. :)
-          xdmp:transaction-locks($host-id, $transaction-id)/host:write[fn:starts-with(., $lock-task-prefix)]/fn:substring-after(., $lock-task-prefix)
-        } catch ($e) {()}
-    )
+    let $transaction-id := xdmp:transaction()
+    let $locked-uris := fn:distinct-values(
+                        for $host-id in xdmp:hosts()
+                        let $check-transactions := xdmp:host-status($host-id)/host:transactions/host:transaction[host:transaction-mode = "update"]/host:transaction-id[. ne $transaction-id]
+                        for $check-transaction in $check-transactions
+                        return xdmp:spawn-function(function() {merge-impl:locked-uris($host-id, $check-transaction)}, map:map() => map:with("update","false") => map:with("priority","higher") => map:with("result",fn:true()))
```
Comment (rjrudin) [inline marklogic-data-hub/src/main/resources/ml-modules/root/com.marklogic.smart-mastering/survivorship/merging/base.xqy]: I think a comment explaining why spawn-function is used here would help as well.

Comment (ryanjdew) [reply marklogic-data-hub/src/main/resources/ml-modules/root/com.marklogic.smart-mastering/survivorship/merging/base.xqy]: Changed to invoke-function and provided a comment in latest push.

File: marklogic-data-hub/src/main/resources/ml-modules/root/com.marklogic.smart-mastering/matcher-impl/helper-impl.xqy [hunk 2]
```diff
@@ -137,3 +137,48 @@ declare function helper-impl:property-name-to-qname($options as element(), $full
         $qname
       )
 };
+
+declare variable $string-token as xs:string := "####";
+
+declare function helper-impl:group-queries-by-scope($queries as cts:query*, $grouping-query-fun as function(cts:query*) as cts:query??) {
```
[the remark this answers is not in the document]
Comment (ryanjdew) [reply marklogic-data-hub/src/main/resources/ml-modules/root/com.marklogic.smart-mastering/matcher-impl/helper-impl.xqy]: comment with explanation added

**Pull request merged**